src/Stratis.Bitcoin/Consensus/ProvenHeadersConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,253 @@
+        /// <seealso cref="https://en.bitcoin.it/wiki/Protocol_documentation#getheaders"/>
+        private const int MaxItemsPerHeadersMessage = 2000;
+
+        public ProvenHeadersConsensusManagerBehavior(ConcurrentChain chain, IInitialBlockDownloadState initialBlockDownloadState, IConsensusManager consensusManager, IPeerBanning peerBanning, ILoggerFactory loggerFactory) : base(chain, initialBlockDownloadState, consensusManager, peerBanning, loggerFactory)
+        {
+            this.chain = chain;
+            this.initialBlockDownloadState = initialBlockDownloadState;
+            this.consensusManager = consensusManager;
+            this.peerBanning = peerBanning;
+            this.loggerFactory = loggerFactory;
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
+        }
+
+        /// <inheritdoc />
+        /// <summary>
+        /// Processes and incoming message from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="message">Received message to process.</param>
+        protected override async Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+        {
+            switch (message.Message.Payload)
+            {
+                case GetHeadersPayload getHeaders:
+                    await this.ProcessGetHeadersAsync(peer, getHeaders).ConfigureAwait(false);
+                    break;
+
+                case HeadersPayload headers:
+                    await this.ProcessHeadersAsync(peer, headers.Headers).ConfigureAwait(false);
+                    break;
+
+                case ProvenHeadersPayload provenHeaders:
+                    await this.ProcessProvenHeadersAsync(peer, provenHeaders).ConfigureAwait(false);
+                    break;
+
+                case GetProvenHeadersPayload getHeaders:
+                    await this.ProcessGetProvenHeadersAsync(peer, getHeaders).ConfigureAwait(false);
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Processes "getprovenheaders" message received from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="getProvenHeadersPayload">Payload of "getprovenheaders" message to process.</param>
+        /// <remarks>
+        /// "getprovenheaders" message is sent by the peer in response to "headers" message until an empty array is received.
+        /// <para>
+        /// This payload notifies peers of our current best validated height, which is held by consensus tip.
+        /// </para>
+        /// <para>
+        /// If the peer is behind/equal to our best height an empty array is sent back.
+        /// </para>
+        /// </remarks>
+        private async Task ProcessGetProvenHeadersAsync(INetworkPeer peer, GetProvenHeadersPayload getProvenHeadersPayload)
```

Comment:
80% of this method is copy paste from `ProcessGetHeadersAsync`

please approach it in similar way like it's done in PoAConsensusManagerBehavior

Reply:
But in PoACMB you don't process "GetPOAHeadersPayload". You using standard GetHeadersPayload.

case GetHeadersPayload getHeaders:
                    await this.ProcessGetHeadersAsync(peer, getHeaders).ConfigureAwait(false);
                    break;